src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -0,0 +1,350 @@
+
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util.{Io, ProgressLogger, Sequences, SimpleCounter}
+import com.fulcrumgenomics.util.Sequences.countMismatches
+import dagr.commons.CommonsDef.{DirPath, PathToBam}
+import dagr.commons.util.LazyLogging
+import dagr.sopt.cmdline.ValidationException
+import dagr.sopt.{arg, clp}
+import htsjdk.samtools.SAMFileHeader.{GroupOrder, SortOrder}
+import htsjdk.samtools._
+import htsjdk.samtools.util.SortingCollection
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+
+object GroupReadsByUmi {
+  private val SortPositionAttribute        = "__GRs"
+  private val LibraryAttribute             = "__GRlb"
+  private val InsertEndsAndStrandAttribute = "__Insert"
+
+  /** Trait that can be implemented to provide a UMI assignment strategy. */
+  sealed trait UmiAssigner {
+    def assign(rawUmis: Seq[String]) : Map[String, String]
+  }
+
+  /** Assigns UMIs based solely on sequence identity. */
+  class IdentityUmiAssigner extends UmiAssigner {
+    override def assign(rawUmis: Seq[String]): Map[String, String] = rawUmis.map(umi => umi -> umi.toUpperCase).toMap
+  }
+
+  /** Assigns UMIs based solely on sequence identity. */
+  class SimpleErrorUmiAssigner(val maxMismatches: Int) extends UmiAssigner {
+    override def assign(rawUmis: Seq[String]): Map[String, String] = {
+      type UmiSet = mutable.SortedSet[String]
+      val umiSets = ListBuffer[UmiSet]()
+      for (umi <- rawUmis) {
+        val matchedSets = mutable.Set[UmiSet]()
+        for (set <- umiSets) {
+          if (set.exists(other => countMismatches(other, umi) <= maxMismatches)) {
+            set.add(umi)
+            matchedSets.add(set)
+          }
+        }
+
+        matchedSets.size match {
+          case 0 => umiSets += mutable.SortedSet(umi)
+          case 1 => matchedSets.head.add(umi)
+          case _ =>
+            // Merge the sets and remove all but one from umiSets
+            val pick = matchedSets.head
+            matchedSets.tail.foreach(set => {
+              umiSets -= set
+              pick ++= set
+            })
+        }
+      }
+
+      umiSets.flatMap(set => set.map(umi => umi -> set.head)).toMap
+    }
+  }
+
+  /** Class that implements the directed adjacency graph method from umi_tools. */
+  class AdjacencyUmiAssigner(val maxMismatches: Int) extends UmiAssigner {
+    /** Represents a node in the adjacency graph; equality is just by UMI sequence. */
+    private class Node(val umi: String, val count: Long, val children: mutable.Buffer[Node] = mutable.Buffer()) {
+      /** Gets the full set of descendants from this node. */
+      def descendants: List[Node] = {
+        val buffer = ListBuffer[Node]()
+        addDescendants(buffer)
+        buffer.toList
+      }
+
+      private def addDescendants(buffer: mutable.Buffer[Node]): Unit = children.foreach(child => {
+        buffer += child
+        child.addDescendants(buffer)
```

Comment:
No - each child is only allowed a single parent in the way I've implemented it.  So I guess really it's an adjacency tree ;)